src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list;
+        list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    List<int> M(IEnumerable<int> nums)
+    {
+        List<int> list;
+        return (from int n1 in nums
+                from int n2 in nums
+                select n1).ToList();
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListSeparateDeclarationAndAssignment()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        List<int> list;
+        list = new List<int>();
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                list.Add(n1);
+            }
+        }|]
+
+        return list.Count();
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        List<int> list;
+        list = (from int n1 in nums
+                from int n2 in nums
+                select n1).ToList();
+        return list.Count();
+    }
+}
+";
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task ToListTypeReplacement01()
+        {
+            string source = @"
+using System;
+using System.Linq;
+using C = System.Collections.Generic.List<int>;
+class Query
+{
+    public static void Main(string[] args)
+    {
+        C c1 = new C { 1, 2, 3 };
+        C c2 = new C { 10, 20, 30 };
+        C c3 = new C { 100, 200, 300 };
+        C r1 = new C();
+        [|foreach (int x in c1)
+        {
+            foreach (int y in c2)
+            {
+                foreach (int z in c3)
+                {
+                    var g = x + y + z;
+                    if (x + y / 10 + z / 100 < 6)
+                    {
+                        r1.Add(g);
+                    }
+                }
+            }
+        }|]
+
+        Console.WriteLine(r1);
+    }
+}
+";
+            string output = @"
+using System;
+using System.Linq;
+using C = System.Collections.Generic.List<int>;
+class Query
+{
+    public static void Main(string[] args)
+    {
+        C c1 = new C { 1, 2, 3 };
+        C c2 = new C { 10, 20, 30 };
+        C c3 = new C { 100, 200, 300 };
+        C r1 = (from int x in c1
```

Comment:
is int there to keep user code? or because it is required? I see some of tests require it since it has type conversion defined. but some seems not needed?

Reply:
This 'int' comes from type in foreach like 'foreach (int x in c1)'

---
In reply to: [195857734](https://github.com/dotnet/roslyn/pull/27130#discussion_r195857734) [](ancestors = 195857734)